python/dgl/distributed/client.py

```diff
@@ -2,33 +2,155 @@
 
 """Functions used by client."""
 
-def connect_to_server(server_namebook):
+from . import rpc
+
+import os
+import time
+import socket
+
+if os.name != 'nt':
+    import fcntl
+    import struct
+
+def local_ip4_addr_list():
+    """Return a set of IPv4 address
+    """
+    nic = set()
+    for if_nidx in socket.if_nameindex():
+        name = if_nidx[1]
+        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        ip_addr = socket.inet_ntoa(fcntl.ioctl(
+            sock.fileno(),
+            0x8915,  # SIOCGIFADDR
+            struct.pack('256s', name[:15].encode("UTF-8")))[20:24])
+        nic.add(ip_addr)
+    return nic
+
+def get_local_machine_id(server_namebook):
+    """Given server_namebook, find local machine ID
+
+    Parameters
+    ----------
+    server_namebook: dict
+        IP address namebook of server nodes, where key is the server's ID
+        (start from 0) and value is the server's machine_id, IP address,
+        port, and group_count, e.g.,
```

Comment:
What is group_count?

Reply:
group_count means how many servers (main_server + bakcup_server) in total inside a machine.